File: ProyectovisualBasic/ProyectovisualBasic/ArbolBinario.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Proyecto_Final1

    Class ArbolBinario

        Private raiz As Nodo

        Private cant As Integer

        Private altura As Integer

        Public Sub New()
            MyBase.New
            Me.raiz = Nothing
        End Sub

        Public Overloads Sub Insertar(ByVal n As Nodo)
            Me.Insertar(Me.raiz, n)
        End Sub

        Public Overloads Sub Insertar(ByRef raiz As Nodo, ByVal n As Nodo)
            If (Me.raiz Is Nothing) Then
                Me.raiz = n
                Return
            ElseIf (n.Dato < Me.raiz.Dato) Then
                Me.Insertar(Me.raiz.izquierdo, n)
            Else
                Me.Insertar(Me.raiz.derecho, n)
            End If

        End Sub

        Private s As String

        Private Overloads Sub PreOrden(ByVal r As Nodo)
            If (Not (r) Is Nothing) Then
                Me.s = (Me.s _
                            + (r.Dato + " "))
                Me.PreOrden(r.izquierdo)
                Me.PreOrden(r.derecho)
            End If

        End Sub

        Public Overloads Function PreOrden() As String
            Me.s = ""
            Me.PreOrden(Me.raiz)
            Return Me.s
        End Function

        Private Overloads Sub InOrden(ByVal r As Nodo)
            If (Not (r) Is Nothing) Then
                Me.InOrden(r.izquierdo)
                Me.s = (Me.s _
                            + (r.Dato + " "))
                Me.InOrden(r.derecho)
            End If

        End Sub

        Public Overloads Function InOrden() As String
            Me.s = ""
            Me.InOrden(Me.raiz)
            Return Me.s
        End Function

        Private Overloads Sub PostOrden(ByVal r As Nodo)
            If (Not (r) Is Nothing) Then
                Me.PostOrden(r.izquierdo)
                Me.PostOrden(r.derecho)
                Me.s = (Me.s _
                            + (r.Dato + " "))
            End If

        End Sub

        Public Overloads Function PostOrden() As String
            Me.s = ""
            Me.PostOrden(Me.raiz)
            Return Me.s
        End Function

        Public Function Buscar(ByVal b As Integer) As Boolean
            Dim r As Nodo = Me.raiz

            While (Not (r) Is Nothing)
                If (b = r.Dato) Then
                    Return True
                End If

                If (r.Dato > b) Then
                    If (r.izquierdo Is Nothing) Then
                        Return False
                    End If

                    r = r.izquierdo
                Else
                    If (r.derecho Is Nothing) Then
                        Return False
                    End If

                    r = r.derecho
                End If


            End While

            Return False
        End Function

        Public Sub Borrar(ByVal b As Integer)
            If (Me.raiz Is Nothing) Then
                Return
            End If

            If Not Me.Buscar(b) Then
                Return
            End If

            Dim r As Nodo = Me.raiz
            Dim Aux As Nodo = Nothing

            While (r.Dato <> b)
                Aux = r
                If (r.Dato > b) Then
                    r = r.izquierdo
                Else
                    r = r.derecho
                End If


            End While

            If (r = Me.raiz) Then
                If (Not (r.derecho) Is Nothing) Then
                    r = r.derecho

                    While (Not (r.izquierdo) Is Nothing)
                        Aux = r
                        r = r.izquierdo

                    End While

                    Me.raiz.Dato = r.Dato
                    Aux.izquierdo = r.derecho
                    Return
                End If

                If (Not (r.izquierdo) Is Nothing) Then
                    r = r.izquierdo
                    Aux = r

                    While (Not (r.derecho) Is Nothing)
                        r = r.derecho

                    End While

                    Me.raiz.Dato = r.Dato
                    Aux.derecho = Aux.derecho.izquierdo
                    Return
                Else
                    Me.raiz = Nothing
                End If

            End If

            If ((r.derecho Is Nothing) _
                        AndAlso (r.izquierdo Is Nothing)) Then
                If (b = Aux.izquierdo.Dato) Then
                    Aux.izquierdo = Nothing
                Else
                    Aux.derecho = Nothing
                End If

            Else
                Dim r2 As Nodo = r
                If (Not (r.derecho) Is Nothing) Then
                    r = r.derecho

                    While (Not (r.izquierdo) Is Nothing)
                        Aux = r
                        r = r.izquierdo

                    End While

                    r2.Dato = r.Dato
                    Aux.izquierdo = Aux.izquierdo.derecho
                    Return
                End If

                If (Not (r.izquierdo) Is Nothing) Then
                    r = r.izquierdo

                    While (Not (r.derecho) Is Nothing)
                        Aux = r
                        r = r.derecho

                    End While

                    r2.Dato = r.Dato
                    Aux.derecho = Aux.derecho.izquierdo
                    Return
                Else
                    Me.raiz = Nothing
                End If

            End If

        End Sub
    End Class
End Namespace

File: Proyecto Riojas/Proyecto Final1/Proyecto Final1/ArbolBinario.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Proyecto_Final1
{
    class ArbolBinario
    {
        private Nodo raiz;
        private int cant;
        private int altura;
        public ArbolBinario()
        {
            raiz = null;
        }

        public void Insertar( Nodo n )
        {
            Insertar(ref raiz, n);
        }
        public void Insertar(ref Nodo raiz, Nodo n)
        {
            if (raiz == null)
            {
                raiz = n;
                return;
            }

            else

            if (n.Dato < raiz.Dato)
                Insertar(ref raiz.izquierdo, n);
                

            else
            {
                Insertar(ref raiz.derecho, n);
            }

        }



        string s ;
        private void PreOrden(Nodo r)
        {
            
            
            if (r != null)
            {
                s += r.Dato + " ";
                PreOrden(r.izquierdo);
                PreOrden (r.derecho);
            }
           
        }
        public string PreOrden()
        {
            s = "";
            PreOrden(raiz);
            return s;
        }

       
        private void InOrden(Nodo r)
        {
            if (r != null)
            {
                InOrden(r.izquierdo);
                s += r.Dato + " ";
                InOrden(r.derecho);

            }
        }

        public string InOrden()
        {
            s = "";
            InOrden(raiz);
            return s;
        }
        
        private void PostOrden(Nodo r)
        {
            if (r != null)
            {
                PostOrden(r.izquierdo);
                PostOrden(r.derecho);
                s += r.Dato + " ";
               
            }
            
        }

        public string PostOrden()
        {
            s = "";
            PostOrden(raiz);
            return s ;
        }


     

        public bool Buscar(int b)
        {
            Nodo r = raiz;


            while (r != null)
            {
                if (b == r.Dato)
                {
                    return true;
                }

                if (r.Dato > b)
                {
                    if (r.izquierdo == null)
                    {
                        return false;
                    }
                    r = r.izquierdo;
                    
                    
                }
               
                else
                {
                    if (r.derecho == null)
                    {
                        return false;
                    }
                    r = r.derecho;
                   
                   
                }
               
            }
           
            return false;
        }

        public void Borrar (int b)
        {
            if (raiz == null)
            {
                return;
            }
            if (!Buscar(b))
            {
                return;
            }
            
            Nodo r = raiz;
            Nodo Aux = null;

            while (r.Dato != b)
            {
                Aux = r;
                if (r.Dato > b)
                    r = r.izquierdo;

                else
                    r = r.derecho;
            }

            if (r == raiz )
            {

                if (r.derecho != null)
                {
                    r = r.derecho;
                    while (r.izquierdo != null)
                    {
                        Aux = r;
                        r = r.izquierdo;
                    }
                    raiz.Dato = r.Dato;
                    Aux.izquierdo = r.derecho;
                    return;
                }
                if (r.izquierdo!= null)
                {
                    r = r.izquierdo;
                    Aux = r;
                    while (r.derecho != null)
                    {
                       
                        r = r.derecho;
                    }
                    raiz.Dato = r.Dato;
                    Aux.derecho = Aux.derecho.izquierdo;
                    return;

                }
                else
                {
                    raiz = null;
                }

            }
            if (r.derecho == null && r.izquierdo == null)
            {
                if (b == Aux.izquierdo.Dato)
                    Aux.izquierdo = null;
                else
                    Aux.derecho = null;

            }
            else
            {
                Nodo r2 = r; 

                if(r.derecho != null)
                {
                    r = r.derecho;
                    while (r.izquierdo != null)
                    {
                        Aux = r;
                        r = r.izquierdo;
                    }
                    r2.Dato = r.Dato;
                    Aux.izquierdo = Aux.izquierdo.derecho;
                    return;
                }
                if (r.izquierdo != null)
                {
                    r = r.izquierdo;
                    while (r.derecho != null)
                    {
                        Aux = r;
                        r = r.derecho;
                    }
                    r2.Dato = r.Dato;
                    Aux.derecho = Aux.derecho.izquierdo;
                    return;

                }
                else
                {
                    raiz = null;
                }
            }




        }
    }
}
